bundles/io/org.eclipse.smarthome.io.net/src/main/java/org/eclipse/smarthome/io/net/http/internal/ExtensibleTrustManager.java

```diff
@@ -0,0 +1,189 @@
+    }
+
+    @Override
+    public void checkServerTrusted(X509Certificate[] chain, String authType, SSLEngine sslEngine)
+            throws CertificateException {
+        X509ExtendedTrustManager linkedTrustManager = getLinkedTrustMananger(chain);
+        if (linkedTrustManager == null) {
+            logger.trace("No specific trust manager found, falling back to default");
+            defaultTrustManager.checkServerTrusted(chain, authType, sslEngine);
+        } else {
+            linkedTrustManager.checkServerTrusted(chain, authType, sslEngine);
+        }
+    }
+
+    private X509ExtendedTrustManager getLinkedTrustMananger(X509Certificate[] chain) {
+        try {
+            String commonName = getCommonName(chain[0]);
+
+            X509ExtendedTrustManager trustManager = linkedTrustManager.getOrDefault(commonName, EMPTY_QUEUE).peek();
+
+            if (trustManager != null) {
+                logger.trace("Found trustManager by common name: {}", commonName);
+                return trustManager;
+            }
+
+            logger.trace("Searching trustManager by Subject Alternative Names: {}",
+                    chain[0].getSubjectAlternativeNames());
+            // @formatter:off
+            return chain[0].getSubjectAlternativeNames().stream()
+                    .map(e -> e.get(1))
+                    .map(Object::toString)
+                    .map(linkedTrustManager::get)
+                    .filter(Objects::nonNull)
+                    .map(Queue::peek)
+                    .filter(Objects::nonNull)
+                    .findFirst()
+                    .orElse(null);
+            // @formatter:on
+        } catch (CertificateParsingException e) {
+            throw new IllegalStateException("Problem while parsing certificate", e);
+        }
+    }
+
+    private String getCommonName(X509Certificate x509Certificate) {
+        String dn = x509Certificate.getSubjectX500Principal().getName(X500Principal.RFC2253);
+        for (String group : dn.split(",")) {
+            if (group.contains("CN=")) {
+                return group.trim().replace("CN=", "");
+            }
+        }
+        throw new IllegalStateException("No Common Name found");
+    }
+
+    @Reference(cardinality = ReferenceCardinality.MULTIPLE, policy = ReferencePolicy.DYNAMIC)
+    protected void addTlsCertificateProvider(TlsCertificateProvider tlsCertificateProvider) {
+        X509ExtendedTrustManager trustManager = new TlsCertificateTrustManagerAdapter(tlsCertificateProvider)
+                .getTrustManager();
+        mappingFromTlsCertificateProvider.put(tlsCertificateProvider, trustManager);
+        addLinkedTrustManager(tlsCertificateProvider.getHostName(), trustManager);
+    }
+
+    protected void removeTlsCertificateProvider(TlsCertificateProvider tlsCertificateProvider) {
+        removeLinkedTrustManager(tlsCertificateProvider.getHostName(),
+                mappingFromTlsCertificateProvider.get(tlsCertificateProvider));
```

Comment:
Would you like to use `get` or `remove`?

Reply:
Great, nicely spotted

Reply:
Pushing it now